lib/tapioca/compilers/gem/listeners/base.rb

```diff
@@ -0,0 +1,50 @@
+# typed: strict
+# frozen_string_literal: true
+
+module Tapioca
+  module Compilers
+    module Gem
+      module Listeners
+        class Base
+          extend T::Sig
+          extend T::Helpers
+
+          abstract!
+
+          sig { params(compiler: SymbolTableCompiler).void }
+          def initialize(compiler)
+            @compiler = compiler
+          end
+
+          sig { params(event: NodeAdded).void }
+          def dispatch(event)
+            case event
+            when ConstNodeAdded
+              on_const(event)
+            when ScopeNodeAdded
+              on_scope(event)
+            when MethodNodeAdded
+              on_method(event)
+            else
+              raise "Unsupported event #{event.class}"
+            end
+          end
+
+          private
+
+          sig { params(event: ConstNodeAdded).void }
+          def on_const(event)
+          end
+
+          sig { params(event: ScopeNodeAdded).void }
+          def on_scope(event)
+          end
+
+          sig { params(event: MethodNodeAdded).void }
+          def on_method(event)
+          end
```

Comment:
Should these be marked as abstract and define them in a single line?

Reply:
I don't think so because if we make them abstract, your listener will have to implement all of them even if it only needs to react on `on_method` for example.

I prefer for them to no-op by default and just override the ones you need.

Especially if we start adding new events like `on_attr`, `on_sig` etc. We'll need to change all the listeners to add an empty override every time we add a new one.